{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from itertools import count\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../data/GPCR_genomic_sequences_Grch37.fasta','r') as fin:\n",
    "    test_header = fin.readline()\n",
    "    test_sequence = fin.readline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'symbol': 'ACKR1',\n",
       " 'name': 'atypical chemokine receptor 1 (Duffy blood group)',\n",
       " 'id': 'ENSG00000213088',\n",
       " 'chr': '1',\n",
       " 'start': 159173097,\n",
       " 'end': 159176290,\n",
       " 'strand': 1}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def parse_header(header):\n",
    "    fields = header.split('|')\n",
    "    return {\n",
    "        'symbol':fields[0][1:],\n",
    "        'name':fields[1],\n",
    "        'id':fields[2],\n",
    "        'chr':fields[3].split('=')[1],\n",
    "        'start':int(fields[4].split('=')[1]),\n",
    "        'end':int(fields[5].split('=')[1]),\n",
    "        'strand':int(fields[6].split('=')[1])\n",
    "    }\n",
    "parse_header(test_header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def variant_array(sequence):\n",
    "    base_array = np.frombuffer(test_sequence.encode('ASCII'),dtype='|S1')[:-1][:,np.newaxis]\n",
    "    seqlength = base_array.shape[0]\n",
    "    condition_list = [\n",
    "        base_array==b'A',\n",
    "        base_array==b'C',\n",
    "        base_array==b'G',\n",
    "        base_array==b'T'\n",
    "    ]\n",
    "    choice_list = [np.tile(x,(seqlength,1)) \\\n",
    "        for x in [\n",
    "            np.array((b'C',b'G',b'T')),\n",
    "            np.array((b'A',b'G',b'T')),\n",
    "            np.array((b'A',b'C',b'T')),\n",
    "            np.array((b'A',b'C',b'G'))\n",
    "        ]\n",
    "    ]\n",
    "    return (base_array, np.select(condition_list,choice_list))\n",
    "\n",
    "test_sequence_array, test_variants_array = variant_array(test_sequence)\n",
    "np.sum(test_sequence_array==test_variants_array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos_start</th>\n",
       "      <th>pos_end</th>\n",
       "      <th>ref/alt</th>\n",
       "      <th>strand</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>159173097</td>\n",
       "      <td>159173097</td>\n",
       "      <td>C/A</td>\n",
       "      <td>1</td>\n",
       "      <td>ACKR1:c.1.C&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>159173097</td>\n",
       "      <td>159173097</td>\n",
       "      <td>C/G</td>\n",
       "      <td>1</td>\n",
       "      <td>ACKR1:c.1.C&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>159173097</td>\n",
       "      <td>159173097</td>\n",
       "      <td>C/T</td>\n",
       "      <td>1</td>\n",
       "      <td>ACKR1:c.1.C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>159173098</td>\n",
       "      <td>159173098</td>\n",
       "      <td>A/C</td>\n",
       "      <td>1</td>\n",
       "      <td>ACKR1:c.2.A&gt;C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>159173098</td>\n",
       "      <td>159173098</td>\n",
       "      <td>A/G</td>\n",
       "      <td>1</td>\n",
       "      <td>ACKR1:c.2.A&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9577</th>\n",
       "      <td>1</td>\n",
       "      <td>159176289</td>\n",
       "      <td>159176289</td>\n",
       "      <td>A/G</td>\n",
       "      <td>1</td>\n",
       "      <td>ACKR1:c.3193.A&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9578</th>\n",
       "      <td>1</td>\n",
       "      <td>159176289</td>\n",
       "      <td>159176289</td>\n",
       "      <td>A/T</td>\n",
       "      <td>1</td>\n",
       "      <td>ACKR1:c.3193.A&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9579</th>\n",
       "      <td>1</td>\n",
       "      <td>159176290</td>\n",
       "      <td>159176290</td>\n",
       "      <td>A/C</td>\n",
       "      <td>1</td>\n",
       "      <td>ACKR1:c.3194.A&gt;C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9580</th>\n",
       "      <td>1</td>\n",
       "      <td>159176290</td>\n",
       "      <td>159176290</td>\n",
       "      <td>A/G</td>\n",
       "      <td>1</td>\n",
       "      <td>ACKR1:c.3194.A&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9581</th>\n",
       "      <td>1</td>\n",
       "      <td>159176290</td>\n",
       "      <td>159176290</td>\n",
       "      <td>A/T</td>\n",
       "      <td>1</td>\n",
       "      <td>ACKR1:c.3194.A&gt;T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9582 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr  pos_start    pos_end ref/alt  strand            symbol\n",
       "0      1  159173097  159173097     C/A       1     ACKR1:c.1.C>A\n",
       "1      1  159173097  159173097     C/G       1     ACKR1:c.1.C>G\n",
       "2      1  159173097  159173097     C/T       1     ACKR1:c.1.C>T\n",
       "3      1  159173098  159173098     A/C       1     ACKR1:c.2.A>C\n",
       "4      1  159173098  159173098     A/G       1     ACKR1:c.2.A>G\n",
       "...   ..        ...        ...     ...     ...               ...\n",
       "9577   1  159176289  159176289     A/G       1  ACKR1:c.3193.A>G\n",
       "9578   1  159176289  159176289     A/T       1  ACKR1:c.3193.A>T\n",
       "9579   1  159176290  159176290     A/C       1  ACKR1:c.3194.A>C\n",
       "9580   1  159176290  159176290     A/G       1  ACKR1:c.3194.A>G\n",
       "9581   1  159176290  159176290     A/T       1  ACKR1:c.3194.A>T\n",
       "\n",
       "[9582 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def synthetic_variants(header, sequence):\n",
    "    # Parse header\n",
    "    header_dict_ = parse_header(header)\n",
    "    # Generate array of alternate bases\n",
    "    sequence_array_, variant_array_ = variant_array(sequence)\n",
    "    # Generate array of positions\n",
    "    pos_array_ = np.arange(len(sequence_array_))\n",
    "    # Tidy to one variant per row\n",
    "    pos = np.repeat(pos_array_,3)\n",
    "    ref = np.repeat(sequence_array_,3).astype(str)\n",
    "    alt = variant_array_.flatten().astype(str)\n",
    "    refalt = np.char.add(np.char.add(ref, '/'), alt)\n",
    "    \n",
    "    # Calculate absolute (assembly) position \n",
    "    if header_dict_['strand'] == 1:\n",
    "        abs_pos = pos + header_dict_['start']\n",
    "    elif header_dict_['strand'] == -1:\n",
    "        abs_pos = - pos + header_dict_['end']\n",
    "    else:\n",
    "        print('Error: Strand unrecognised!')\n",
    "    # Generate ids\n",
    "    symbol = np.char.add(\n",
    "        np.char.add(\n",
    "            header_dict_['symbol'] + ':c.',\n",
    "            (pos + 1).astype(str)\n",
    "        ),\n",
    "        np.char.add(\n",
    "            np.char.add('.', ref),\n",
    "            np.char.add('>', alt)\n",
    "        )\n",
    "    )\n",
    "    \n",
    "    # Sanitise for Ensembl VEP\n",
    "    output = pd.DataFrame(np.stack((abs_pos,abs_pos,refalt,symbol),axis=1),\n",
    "                          columns = ['pos_start','pos_end','ref/alt','symbol'])\n",
    "    # Add chromosome\n",
    "    output['chr'] = header_dict_['chr']\n",
    "    # Add strand\n",
    "    output['strand'] = header_dict_['strand']\n",
    "    # Reorder columns\n",
    "    output = output[['chr','pos_start','pos_end','ref/alt','strand','symbol']]\n",
    "    return output\n",
    "    \n",
    "synthetic_variants(test_header,test_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Synthetic variant file wiped.\n"
     ]
    }
   ],
   "source": [
    "with open('../data/GPCR_synthetic_variants.txt','w') as fid:\n",
    "    print('Synthetic variant file wiped.')\n",
    "with open('../data/GPCR_genomic_sequences_Grch37.fasta','r') as fin:\n",
    "    while True:\n",
    "        header = fin.readline()\n",
    "        sequence = fin.readline()\n",
    "        if not sequence: \n",
    "            break\n",
    "        syn_var_ = synthetic_variants(header, sequence)\n",
    "        syn_var_.to_csv('../data/GPCR_synthetic_variants.txt',sep='\\t',mode='a',index=False,header=False)\n",
    "        "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:variants]",
   "language": "python",
   "name": "conda-env-variants-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
