{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC symbol</th>\n",
       "      <th>HGNC name</th>\n",
       "      <th>HGNC id</th>\n",
       "      <th>Ensembl id Grch38</th>\n",
       "      <th>Grch37 symbol</th>\n",
       "      <th>Ensembl id Grch37</th>\n",
       "      <th>Grch37 chromosome</th>\n",
       "      <th>Grch37 start bp</th>\n",
       "      <th>Grch37 end bp</th>\n",
       "      <th>Grch37 strand</th>\n",
       "      <th>Human genetic localisation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACKR1</td>\n",
       "      <td>atypical chemokine receptor 1 (Duffy blood group)</td>\n",
       "      <td>4035</td>\n",
       "      <td>ENSG00000213088</td>\n",
       "      <td>DARC</td>\n",
       "      <td>ENSG00000213088</td>\n",
       "      <td>1</td>\n",
       "      <td>159173097</td>\n",
       "      <td>159176290</td>\n",
       "      <td>1</td>\n",
       "      <td>1q23.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACKR2</td>\n",
       "      <td>atypical chemokine receptor 2</td>\n",
       "      <td>1565</td>\n",
       "      <td>ENSG00000144648</td>\n",
       "      <td>ACKR2</td>\n",
       "      <td>ENSG00000144648</td>\n",
       "      <td>3</td>\n",
       "      <td>42846244</td>\n",
       "      <td>42929466</td>\n",
       "      <td>1</td>\n",
       "      <td>3p22.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACKR3</td>\n",
       "      <td>atypical chemokine receptor 3</td>\n",
       "      <td>23692</td>\n",
       "      <td>ENSG00000144476</td>\n",
       "      <td>ACKR3</td>\n",
       "      <td>ENSG00000144476</td>\n",
       "      <td>2</td>\n",
       "      <td>237476430</td>\n",
       "      <td>237491001</td>\n",
       "      <td>1</td>\n",
       "      <td>2q37.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACKR4</td>\n",
       "      <td>atypical chemokine receptor 4</td>\n",
       "      <td>1611</td>\n",
       "      <td>ENSG00000129048</td>\n",
       "      <td>ACKR4</td>\n",
       "      <td>ENSG00000129048</td>\n",
       "      <td>3</td>\n",
       "      <td>132316081</td>\n",
       "      <td>132337811</td>\n",
       "      <td>1</td>\n",
       "      <td>3q22.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADCYAP1R1</td>\n",
       "      <td>ADCYAP receptor type I</td>\n",
       "      <td>242</td>\n",
       "      <td>ENSG00000078549</td>\n",
       "      <td>ADCYAP1R1</td>\n",
       "      <td>ENSG00000078549</td>\n",
       "      <td>7</td>\n",
       "      <td>31092076</td>\n",
       "      <td>31151089</td>\n",
       "      <td>1</td>\n",
       "      <td>7p14.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388</th>\n",
       "      <td>TSHR</td>\n",
       "      <td>thyroid stimulating hormone receptor</td>\n",
       "      <td>12373</td>\n",
       "      <td>ENSG00000165409</td>\n",
       "      <td>TSHR</td>\n",
       "      <td>ENSG00000165409</td>\n",
       "      <td>14</td>\n",
       "      <td>81421333</td>\n",
       "      <td>81612646</td>\n",
       "      <td>1</td>\n",
       "      <td>14q24-q31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389</th>\n",
       "      <td>UTS2R</td>\n",
       "      <td>urotensin 2 receptor</td>\n",
       "      <td>4468</td>\n",
       "      <td>ENSG00000181408</td>\n",
       "      <td>UTS2R</td>\n",
       "      <td>ENSG00000181408</td>\n",
       "      <td>17</td>\n",
       "      <td>80332153</td>\n",
       "      <td>80333462</td>\n",
       "      <td>1</td>\n",
       "      <td>17q25.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>VIPR1</td>\n",
       "      <td>vasoactive intestinal peptide receptor 1</td>\n",
       "      <td>12694</td>\n",
       "      <td>ENSG00000114812</td>\n",
       "      <td>VIPR1</td>\n",
       "      <td>ENSG00000114812</td>\n",
       "      <td>3</td>\n",
       "      <td>42530791</td>\n",
       "      <td>42579059</td>\n",
       "      <td>1</td>\n",
       "      <td>3p22.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391</th>\n",
       "      <td>VIPR2</td>\n",
       "      <td>vasoactive intestinal peptide receptor 2</td>\n",
       "      <td>12695</td>\n",
       "      <td>ENSG00000106018</td>\n",
       "      <td>VIPR2</td>\n",
       "      <td>ENSG00000106018</td>\n",
       "      <td>7</td>\n",
       "      <td>158820866</td>\n",
       "      <td>158937649</td>\n",
       "      <td>-1</td>\n",
       "      <td>7q36.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>XCR1</td>\n",
       "      <td>X-C motif chemokine receptor 1</td>\n",
       "      <td>1625</td>\n",
       "      <td>ENSG00000173578</td>\n",
       "      <td>XCR1</td>\n",
       "      <td>ENSG00000173578</td>\n",
       "      <td>3</td>\n",
       "      <td>46058516</td>\n",
       "      <td>46069234</td>\n",
       "      <td>-1</td>\n",
       "      <td>3p21.31</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>393 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    HGNC symbol                                          HGNC name  HGNC id  \\\n",
       "0         ACKR1  atypical chemokine receptor 1 (Duffy blood group)     4035   \n",
       "1         ACKR2                      atypical chemokine receptor 2     1565   \n",
       "2         ACKR3                      atypical chemokine receptor 3    23692   \n",
       "3         ACKR4                      atypical chemokine receptor 4     1611   \n",
       "4     ADCYAP1R1                             ADCYAP receptor type I      242   \n",
       "..          ...                                                ...      ...   \n",
       "388        TSHR               thyroid stimulating hormone receptor    12373   \n",
       "389       UTS2R                               urotensin 2 receptor     4468   \n",
       "390       VIPR1           vasoactive intestinal peptide receptor 1    12694   \n",
       "391       VIPR2           vasoactive intestinal peptide receptor 2    12695   \n",
       "392        XCR1                     X-C motif chemokine receptor 1     1625   \n",
       "\n",
       "    Ensembl id Grch38 Grch37 symbol Ensembl id Grch37 Grch37 chromosome  \\\n",
       "0     ENSG00000213088          DARC   ENSG00000213088                 1   \n",
       "1     ENSG00000144648         ACKR2   ENSG00000144648                 3   \n",
       "2     ENSG00000144476         ACKR3   ENSG00000144476                 2   \n",
       "3     ENSG00000129048         ACKR4   ENSG00000129048                 3   \n",
       "4     ENSG00000078549     ADCYAP1R1   ENSG00000078549                 7   \n",
       "..                ...           ...               ...               ...   \n",
       "388   ENSG00000165409          TSHR   ENSG00000165409                14   \n",
       "389   ENSG00000181408         UTS2R   ENSG00000181408                17   \n",
       "390   ENSG00000114812         VIPR1   ENSG00000114812                 3   \n",
       "391   ENSG00000106018         VIPR2   ENSG00000106018                 7   \n",
       "392   ENSG00000173578          XCR1   ENSG00000173578                 3   \n",
       "\n",
       "     Grch37 start bp  Grch37 end bp  Grch37 strand Human genetic localisation  \n",
       "0          159173097      159176290              1                     1q23.2  \n",
       "1           42846244       42929466              1                     3p22.1  \n",
       "2          237476430      237491001              1                     2q37.3  \n",
       "3          132316081      132337811              1                     3q22.1  \n",
       "4           31092076       31151089              1                     7p14.3  \n",
       "..               ...            ...            ...                        ...  \n",
       "388         81421333       81612646              1                  14q24-q31  \n",
       "389         80332153       80333462              1                    17q25.3  \n",
       "390         42530791       42579059              1                     3p22.1  \n",
       "391        158820866      158937649             -1                     7q36.3  \n",
       "392         46058516       46069234             -1                    3p21.31  \n",
       "\n",
       "[393 rows x 11 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load table of GPCR gene regions, extracted from Ensembl\n",
    "gpcr_gene_regions = pd.read_csv('../data/Ensembl_Grch37_gpcr_genome_locations.csv')\n",
    "gpcr_gene_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC Symbol</th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>Ensembl ID</th>\n",
       "      <th>Class</th>\n",
       "      <th>Ligand type</th>\n",
       "      <th>Family</th>\n",
       "      <th>Gs</th>\n",
       "      <th>Gi/o</th>\n",
       "      <th>Gq/11</th>\n",
       "      <th>G12/13</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCAR1</td>\n",
       "      <td>HCAR1_HUMAN</td>\n",
       "      <td>ENSG00000196917</td>\n",
       "      <td>A</td>\n",
       "      <td>Alicarboxylic acid</td>\n",
       "      <td>Hydroxycarboxylic acid</td>\n",
       "      <td>-</td>\n",
       "      <td>Primary</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCAR2</td>\n",
       "      <td>HCAR2_HUMAN</td>\n",
       "      <td>ENSG00000182782</td>\n",
       "      <td>A</td>\n",
       "      <td>Alicarboxylic acid</td>\n",
       "      <td>Hydroxycarboxylic acid</td>\n",
       "      <td>-</td>\n",
       "      <td>Primary</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCAR3</td>\n",
       "      <td>HCAR3_HUMAN</td>\n",
       "      <td>ENSG00000255398</td>\n",
       "      <td>A</td>\n",
       "      <td>Alicarboxylic acid</td>\n",
       "      <td>Hydroxycarboxylic acid</td>\n",
       "      <td>-</td>\n",
       "      <td>Primary</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OXGR1</td>\n",
       "      <td>OXGR1_HUMAN</td>\n",
       "      <td>ENSG00000165621</td>\n",
       "      <td>A</td>\n",
       "      <td>Alicarboxylic acid</td>\n",
       "      <td>Oxoglutarate</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Primary</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SUCNR1</td>\n",
       "      <td>SUCR1_HUMAN</td>\n",
       "      <td>ENSG00000198829</td>\n",
       "      <td>A</td>\n",
       "      <td>Alicarboxylic acid</td>\n",
       "      <td>Succinate</td>\n",
       "      <td>-</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Secondary</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>TAS2R4</td>\n",
       "      <td>TA2R4_HUMAN</td>\n",
       "      <td>ENSG00000127364</td>\n",
       "      <td>T</td>\n",
       "      <td>Sensory</td>\n",
       "      <td>Taste 2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393</th>\n",
       "      <td>TAS2R5</td>\n",
       "      <td>TA2R5_HUMAN</td>\n",
       "      <td>ENSG00000127366</td>\n",
       "      <td>T</td>\n",
       "      <td>Sensory</td>\n",
       "      <td>Taste 2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394</th>\n",
       "      <td>TAS2R7</td>\n",
       "      <td>TA2R7_HUMAN</td>\n",
       "      <td>ENSG00000121377</td>\n",
       "      <td>T</td>\n",
       "      <td>Sensory</td>\n",
       "      <td>Taste 2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>TAS2R8</td>\n",
       "      <td>TA2R8_HUMAN</td>\n",
       "      <td>ENSG00000121314</td>\n",
       "      <td>T</td>\n",
       "      <td>Sensory</td>\n",
       "      <td>Taste 2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>TAS2R9</td>\n",
       "      <td>TA2R9_HUMAN</td>\n",
       "      <td>ENSG00000121381</td>\n",
       "      <td>T</td>\n",
       "      <td>Sensory</td>\n",
       "      <td>Taste 2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>397 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    HGNC Symbol      Uniprot       Ensembl ID Class         Ligand type  \\\n",
       "0         HCAR1  HCAR1_HUMAN  ENSG00000196917     A  Alicarboxylic acid   \n",
       "1         HCAR2  HCAR2_HUMAN  ENSG00000182782     A  Alicarboxylic acid   \n",
       "2         HCAR3  HCAR3_HUMAN  ENSG00000255398     A  Alicarboxylic acid   \n",
       "3         OXGR1  OXGR1_HUMAN  ENSG00000165621     A  Alicarboxylic acid   \n",
       "4        SUCNR1  SUCR1_HUMAN  ENSG00000198829     A  Alicarboxylic acid   \n",
       "..          ...          ...              ...   ...                 ...   \n",
       "392      TAS2R4  TA2R4_HUMAN  ENSG00000127364     T             Sensory   \n",
       "393      TAS2R5  TA2R5_HUMAN  ENSG00000127366     T             Sensory   \n",
       "394      TAS2R7  TA2R7_HUMAN  ENSG00000121377     T             Sensory   \n",
       "395      TAS2R8  TA2R8_HUMAN  ENSG00000121314     T             Sensory   \n",
       "396      TAS2R9  TA2R9_HUMAN  ENSG00000121381     T             Sensory   \n",
       "\n",
       "                     Family Gs     Gi/o      Gq/11 G12/13  \n",
       "0    Hydroxycarboxylic acid  -  Primary          -      -  \n",
       "1    Hydroxycarboxylic acid  -  Primary          -      -  \n",
       "2    Hydroxycarboxylic acid  -  Primary          -      -  \n",
       "3              Oxoglutarate  -        -    Primary      -  \n",
       "4                 Succinate  -  Primary  Secondary      -  \n",
       "..                      ... ..      ...        ...    ...  \n",
       "392                 Taste 2  -        -          -      -  \n",
       "393                 Taste 2  -        -          -      -  \n",
       "394                 Taste 2  -        -          -      -  \n",
       "395                 Taste 2  -        -          -      -  \n",
       "396                 Taste 2  -        -          -      -  \n",
       "\n",
       "[397 rows x 10 columns]"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gpcrs = pd.read_csv('../data/GPCRdb_gpcr_classes.csv')\n",
    "gpcrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Orphan                80\n",
       "Peptide               75\n",
       "Aminergic             36\n",
       "Lipid                 36\n",
       "Protein               29\n",
       "Nucleotide            12\n",
       "Sensory                7\n",
       "Alicarboxylic acid     5\n",
       "Olfactory              3\n",
       "Melatonin              2\n",
       "Steroid                2\n",
       "Name: Ligand type, dtype: int64"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gpcrs[gpcrs.Class=='A']['Ligand type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>...</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>NumberSubmitters</th>\n",
       "      <th>Guidelines</th>\n",
       "      <th>TestedInGTR</th>\n",
       "      <th>OtherIDs</th>\n",
       "      <th>SubmitterCategories</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15041</td>\n",
       "      <td>Indel</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704705</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215070,OMIM:613653.0001</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4820844</td>\n",
       "      <td>GGAT</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15041</td>\n",
       "      <td>Indel</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704705</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215070,OMIM:613653.0001</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4781213</td>\n",
       "      <td>GGAT</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15042</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704709</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215072,OMIM:613653.0002</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4827360</td>\n",
       "      <td>GCTGCTGGACCTGCC</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15042</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704709</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215072,OMIM:613653.0002</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4787729</td>\n",
       "      <td>GCTGCTGGACCTGCC</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15043</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_014630.3(ZNF592):c.3136G&gt;A (p.Gly1046Arg)</td>\n",
       "      <td>9640</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>HGNC:28986</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>0</td>\n",
       "      <td>Jun 29, 2015</td>\n",
       "      <td>150829393</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA210674,UniProtKB:Q92610#VAR_064583,O...</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>85342440</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID                       Type  \\\n",
       "0      15041                      Indel   \n",
       "1      15041                      Indel   \n",
       "2      15042                   Deletion   \n",
       "3      15042                   Deletion   \n",
       "4      15043  single nucleotide variant   \n",
       "\n",
       "                                                Name  GeneID GeneSymbol  \\\n",
       "0  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...    9907      AP5Z1   \n",
       "1  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...    9907      AP5Z1   \n",
       "2     NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)    9907      AP5Z1   \n",
       "3     NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)    9907      AP5Z1   \n",
       "4       NM_014630.3(ZNF592):c.3136G>A (p.Gly1046Arg)    9640     ZNF592   \n",
       "\n",
       "      HGNC_ID    ClinicalSignificance  ClinSigSimple LastEvaluated  \\\n",
       "0  HGNC:22197              Pathogenic              1  Jun 29, 2010   \n",
       "1  HGNC:22197              Pathogenic              1  Jun 29, 2010   \n",
       "2  HGNC:22197              Pathogenic              1  Jun 29, 2010   \n",
       "3  HGNC:22197              Pathogenic              1  Jun 29, 2010   \n",
       "4  HGNC:28986  Uncertain significance              0  Jun 29, 2015   \n",
       "\n",
       "   RS# (dbSNP)  ...                    ReviewStatus NumberSubmitters  \\\n",
       "0    397704705  ...  no assertion criteria provided                1   \n",
       "1    397704705  ...  no assertion criteria provided                1   \n",
       "2    397704709  ...  no assertion criteria provided                1   \n",
       "3    397704709  ...  no assertion criteria provided                1   \n",
       "4    150829393  ...  no assertion criteria provided                1   \n",
       "\n",
       "  Guidelines TestedInGTR                                           OtherIDs  \\\n",
       "0          -           N                  ClinGen:CA215070,OMIM:613653.0001   \n",
       "1          -           N                  ClinGen:CA215070,OMIM:613653.0001   \n",
       "2          -           N                  ClinGen:CA215072,OMIM:613653.0002   \n",
       "3          -           N                  ClinGen:CA215072,OMIM:613653.0002   \n",
       "4          -           N  ClinGen:CA210674,UniProtKB:Q92610#VAR_064583,O...   \n",
       "\n",
       "  SubmitterCategories VariationID PositionVCF ReferenceAlleleVCF  \\\n",
       "0                   1           2     4820844               GGAT   \n",
       "1                   1           2     4781213               GGAT   \n",
       "2                   1           3     4827360    GCTGCTGGACCTGCC   \n",
       "3                   1           3     4787729    GCTGCTGGACCTGCC   \n",
       "4                   1           4    85342440                  G   \n",
       "\n",
       "       AlternateAlleleVCF  \n",
       "0  TGCTGTAAACTGTAACTGTAAA  \n",
       "1  TGCTGTAAACTGTAACTGTAAA  \n",
       "2                       G  \n",
       "3                       G  \n",
       "4                       A  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_variants = pd.read_csv('../data/variant_summary.txt',sep='\\t',low_memory=False)\n",
    "clinvar_variants.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "#AlleleID                                                           15041\n",
       "Type                                                                Indel\n",
       "Name                    NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...\n",
       "GeneID                                                               9907\n",
       "GeneSymbol                                                          AP5Z1\n",
       "HGNC_ID                                                        HGNC:22197\n",
       "ClinicalSignificance                                           Pathogenic\n",
       "ClinSigSimple                                                           1\n",
       "LastEvaluated                                                Jun 29, 2010\n",
       "RS# (dbSNP)                                                     397704705\n",
       "nsv/esv (dbVar)                                                         -\n",
       "RCVaccession                                                 RCV000000012\n",
       "PhenotypeIDS            MONDO:MONDO:0013342,MedGen:C3150901,OMIM:61364...\n",
       "PhenotypeList                  Spastic paraplegia 48, autosomal recessive\n",
       "Origin                                                           germline\n",
       "OriginSimple                                                     germline\n",
       "Assembly                                                           GRCh37\n",
       "ChromosomeAccession                                          NC_000007.13\n",
       "Chromosome                                                              7\n",
       "Start                                                             4820844\n",
       "Stop                                                              4820847\n",
       "ReferenceAllele                                                        na\n",
       "AlternateAllele                                                        na\n",
       "Cytogenetic                                                        7p22.1\n",
       "ReviewStatus                               no assertion criteria provided\n",
       "NumberSubmitters                                                        1\n",
       "Guidelines                                                              -\n",
       "TestedInGTR                                                             N\n",
       "OtherIDs                                ClinGen:CA215070,OMIM:613653.0001\n",
       "SubmitterCategories                                                     1\n",
       "VariationID                                                             2\n",
       "PositionVCF                                                       4820844\n",
       "ReferenceAlleleVCF                                                   GGAT\n",
       "AlternateAlleleVCF                                 TGCTGTAAACTGTAACTGTAAA\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_variants.iloc[0,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "#AlleleID                                                           15041\n",
       "Type                                                                Indel\n",
       "Name                    NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...\n",
       "GeneID                                                               9907\n",
       "GeneSymbol                                                          AP5Z1\n",
       "HGNC_ID                                                        HGNC:22197\n",
       "ClinicalSignificance                                           Pathogenic\n",
       "ClinSigSimple                                                           1\n",
       "LastEvaluated                                                Jun 29, 2010\n",
       "RS# (dbSNP)                                                     397704705\n",
       "nsv/esv (dbVar)                                                         -\n",
       "RCVaccession                                                 RCV000000012\n",
       "PhenotypeIDS            MONDO:MONDO:0013342,MedGen:C3150901,OMIM:61364...\n",
       "PhenotypeList                  Spastic paraplegia 48, autosomal recessive\n",
       "Origin                                                           germline\n",
       "OriginSimple                                                     germline\n",
       "Assembly                                                           GRCh38\n",
       "ChromosomeAccession                                          NC_000007.14\n",
       "Chromosome                                                              7\n",
       "Start                                                             4781213\n",
       "Stop                                                              4781216\n",
       "ReferenceAllele                                                        na\n",
       "AlternateAllele                                                        na\n",
       "Cytogenetic                                                        7p22.1\n",
       "ReviewStatus                               no assertion criteria provided\n",
       "NumberSubmitters                                                        1\n",
       "Guidelines                                                              -\n",
       "TestedInGTR                                                             N\n",
       "OtherIDs                                ClinGen:CA215070,OMIM:613653.0001\n",
       "SubmitterCategories                                                     1\n",
       "VariationID                                                             2\n",
       "PositionVCF                                                       4781213\n",
       "ReferenceAlleleVCF                                                   GGAT\n",
       "AlternateAlleleVCF                                 TGCTGTAAACTGTAACTGTAAA\n",
       "Name: 1, dtype: object"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_variants.iloc[1,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC Symbol</th>\n",
       "      <th>Class</th>\n",
       "      <th>Name</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>PhenotypeIDS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCAR2</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh37/hg19 12q24.31(chr12:123185916-123187546)x3</td>\n",
       "      <td>Benign</td>\n",
       "      <td>See cases</td>\n",
       "      <td>not provided</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>12</td>\n",
       "      <td>-1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCAR2</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_177551.4(HCAR2):c.951G&gt;A (p.Met317Ile)</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>12</td>\n",
       "      <td>123186880</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2454727</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCAR2</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_177551.4(HCAR2):c.951G&gt;A (p.Met317Ile)</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>12</td>\n",
       "      <td>122702333</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2454727</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OXGR1</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_080818.4(OXGR1):c.134del (p.Leu45fs)</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>13</td>\n",
       "      <td>97639879</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "      <td>145685926</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OXGR1</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_080818.4(OXGR1):c.134del (p.Leu45fs)</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>13</td>\n",
       "      <td>96987625</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "      <td>145685926</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16627</th>\n",
       "      <td>TAS2R9</td>\n",
       "      <td>T</td>\n",
       "      <td>NM_023917.2(TAS2R9):c.653G&gt;A (p.Gly218Glu)</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>12</td>\n",
       "      <td>10809423</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>113883583</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16628</th>\n",
       "      <td>TAS2R9</td>\n",
       "      <td>T</td>\n",
       "      <td>NM_023917.2(TAS2R9):c.450T&gt;A (p.Asp150Glu)</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>12</td>\n",
       "      <td>10962225</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>74954631</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16629</th>\n",
       "      <td>TAS2R9</td>\n",
       "      <td>T</td>\n",
       "      <td>NM_023917.2(TAS2R9):c.450T&gt;A (p.Asp150Glu)</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>12</td>\n",
       "      <td>10809626</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>74954631</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16630</th>\n",
       "      <td>TAS2R9</td>\n",
       "      <td>T</td>\n",
       "      <td>NM_023917.2(TAS2R9):c.201C&gt;A (p.Phe67Leu)</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>12</td>\n",
       "      <td>10962474</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>61735347</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16631</th>\n",
       "      <td>TAS2R9</td>\n",
       "      <td>T</td>\n",
       "      <td>NM_023917.2(TAS2R9):c.201C&gt;A (p.Phe67Leu)</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>12</td>\n",
       "      <td>10809875</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>61735347</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16632 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      HGNC Symbol Class                                               Name  \\\n",
       "0           HCAR2     A  GRCh37/hg19 12q24.31(chr12:123185916-123187546)x3   \n",
       "1           HCAR2     A          NM_177551.4(HCAR2):c.951G>A (p.Met317Ile)   \n",
       "2           HCAR2     A          NM_177551.4(HCAR2):c.951G>A (p.Met317Ile)   \n",
       "3           OXGR1     A            NM_080818.4(OXGR1):c.134del (p.Leu45fs)   \n",
       "4           OXGR1     A            NM_080818.4(OXGR1):c.134del (p.Leu45fs)   \n",
       "...           ...   ...                                                ...   \n",
       "16627      TAS2R9     T         NM_023917.2(TAS2R9):c.653G>A (p.Gly218Glu)   \n",
       "16628      TAS2R9     T         NM_023917.2(TAS2R9):c.450T>A (p.Asp150Glu)   \n",
       "16629      TAS2R9     T         NM_023917.2(TAS2R9):c.450T>A (p.Asp150Glu)   \n",
       "16630      TAS2R9     T          NM_023917.2(TAS2R9):c.201C>A (p.Phe67Leu)   \n",
       "16631      TAS2R9     T          NM_023917.2(TAS2R9):c.201C>A (p.Phe67Leu)   \n",
       "\n",
       "      ClinicalSignificance PhenotypeList  OriginSimple  \\\n",
       "0                   Benign     See cases  not provided   \n",
       "1                   Benign  not provided      germline   \n",
       "2                   Benign  not provided      germline   \n",
       "3                   Benign  not provided      germline   \n",
       "4                   Benign  not provided      germline   \n",
       "...                    ...           ...           ...   \n",
       "16627               Benign  not provided      germline   \n",
       "16628               Benign  not provided      germline   \n",
       "16629               Benign  not provided      germline   \n",
       "16630               Benign  not provided      germline   \n",
       "16631               Benign  not provided      germline   \n",
       "\n",
       "                              ReviewStatus Assembly Chromosome  PositionVCF  \\\n",
       "0           no assertion criteria provided   GRCh37         12           -1   \n",
       "1      criteria provided, single submitter   GRCh37         12    123186880   \n",
       "2      criteria provided, single submitter   GRCh38         12    122702333   \n",
       "3      criteria provided, single submitter   GRCh37         13     97639879   \n",
       "4      criteria provided, single submitter   GRCh38         13     96987625   \n",
       "...                                    ...      ...        ...          ...   \n",
       "16627  criteria provided, single submitter   GRCh38         12     10809423   \n",
       "16628  criteria provided, single submitter   GRCh37         12     10962225   \n",
       "16629  criteria provided, single submitter   GRCh38         12     10809626   \n",
       "16630  criteria provided, single submitter   GRCh37         12     10962474   \n",
       "16631  criteria provided, single submitter   GRCh38         12     10809875   \n",
       "\n",
       "      ReferenceAlleleVCF AlternateAlleleVCF  RS# (dbSNP)     PhenotypeIDS  \n",
       "0                     na                 na           -1                -  \n",
       "1                      C                  T      2454727  MedGen:CN517202  \n",
       "2                      C                  T      2454727  MedGen:CN517202  \n",
       "3                     GA                  G    145685926  MedGen:CN517202  \n",
       "4                     GA                  G    145685926  MedGen:CN517202  \n",
       "...                  ...                ...          ...              ...  \n",
       "16627                  C                  T    113883583  MedGen:CN517202  \n",
       "16628                  A                  T     74954631  MedGen:CN517202  \n",
       "16629                  A                  T     74954631  MedGen:CN517202  \n",
       "16630                  G                  T     61735347  MedGen:CN517202  \n",
       "16631                  G                  T     61735347  MedGen:CN517202  \n",
       "\n",
       "[16632 rows x 14 columns]"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_gpcr_variants = gpcrs.merge(clinvar_variants,left_on='HGNC Symbol',right_on='GeneSymbol')\n",
    "clinvar_gpcr_variants = clinvar_gpcr_variants[[\n",
    "    'HGNC Symbol',\n",
    "    'Class',\n",
    "    'Name',\n",
    "    'ClinicalSignificance',\n",
    "    'PhenotypeList',\n",
    "    'OriginSimple',\n",
    "    'ReviewStatus',\n",
    "    'Assembly',\n",
    "    'Chromosome',\n",
    "    'PositionVCF',\n",
    "    'ReferenceAlleleVCF',\n",
    "    'AlternateAlleleVCF',\n",
    "    'RS# (dbSNP)',\n",
    "    'PhenotypeIDS'\n",
    "]]\n",
    "clinvar_gpcr_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC Symbol</th>\n",
       "      <th>Class</th>\n",
       "      <th>Name</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>PhenotypeIDS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCAR2</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_177551.4(HCAR2):c.951G&gt;A (p.Met317Ile)</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>12</td>\n",
       "      <td>123186880</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2454727</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OXGR1</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_080818.4(OXGR1):c.134del (p.Leu45fs)</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>13</td>\n",
       "      <td>97639879</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "      <td>145685926</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SUCNR1</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_033050.6(SUCNR1):c.195C&gt;T (p.Leu65=)</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>3</td>\n",
       "      <td>151598526</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>79888638</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HTR1A</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_000524.3(HTR1A):c.-1019G&gt;G</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not specified</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>5</td>\n",
       "      <td>63258565</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>6295</td>\n",
       "      <td>MedGen:CN169374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HTR1A</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_000524.3(HTR1A):c.-1019G&gt;C</td>\n",
       "      <td>drug response</td>\n",
       "      <td>paroxetine response - Efficacy|not specified</td>\n",
       "      <td>germline</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>5</td>\n",
       "      <td>63258565</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>6295</td>\n",
       "      <td>MedGen:CN236573|MedGen:CN169374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7332</th>\n",
       "      <td>GPBAR1</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_170699.3(GPBAR1):c.845del (p.Leu282fs)</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>2</td>\n",
       "      <td>219128291</td>\n",
       "      <td>CT</td>\n",
       "      <td>C</td>\n",
       "      <td>775667051</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7334</th>\n",
       "      <td>GPBAR1</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_170699.3(GPBAR1):c.887A&gt;G (p.Gln296Arg)</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>2</td>\n",
       "      <td>219128334</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>115448647</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7336</th>\n",
       "      <td>GPBAR1</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_170699.3(GPBAR1):c.21C&gt;G (p.Gly7=)</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>2</td>\n",
       "      <td>219127468</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>201145373</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7338</th>\n",
       "      <td>GPBAR1</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_170699.3(GPBAR1):c.117C&gt;T (p.Ile39=)</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>2</td>\n",
       "      <td>219127564</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>770486496</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7340</th>\n",
       "      <td>GPER1</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_001098201.3(GPER1):c.14C&gt;T (p.Ser5Phe)</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7</td>\n",
       "      <td>1131378</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>117290655</td>\n",
       "      <td>MedGen:CN517202</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2760 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     HGNC Symbol Class                                        Name  \\\n",
       "1          HCAR2     A   NM_177551.4(HCAR2):c.951G>A (p.Met317Ile)   \n",
       "3          OXGR1     A     NM_080818.4(OXGR1):c.134del (p.Leu45fs)   \n",
       "5         SUCNR1     A     NM_033050.6(SUCNR1):c.195C>T (p.Leu65=)   \n",
       "9          HTR1A     A               NM_000524.3(HTR1A):c.-1019G>G   \n",
       "11         HTR1A     A               NM_000524.3(HTR1A):c.-1019G>C   \n",
       "...          ...   ...                                         ...   \n",
       "7332      GPBAR1     A   NM_170699.3(GPBAR1):c.845del (p.Leu282fs)   \n",
       "7334      GPBAR1     A  NM_170699.3(GPBAR1):c.887A>G (p.Gln296Arg)   \n",
       "7336      GPBAR1     A       NM_170699.3(GPBAR1):c.21C>G (p.Gly7=)   \n",
       "7338      GPBAR1     A     NM_170699.3(GPBAR1):c.117C>T (p.Ile39=)   \n",
       "7340       GPER1     A   NM_001098201.3(GPER1):c.14C>T (p.Ser5Phe)   \n",
       "\n",
       "        ClinicalSignificance                                 PhenotypeList  \\\n",
       "1                     Benign                                  not provided   \n",
       "3                     Benign                                  not provided   \n",
       "5                     Benign                                  not provided   \n",
       "9                     Benign                                 not specified   \n",
       "11             drug response  paroxetine response - Efficacy|not specified   \n",
       "...                      ...                                           ...   \n",
       "7332  Uncertain significance                                  not provided   \n",
       "7334                  Benign                                  not provided   \n",
       "7336                  Benign                                  not provided   \n",
       "7338           Likely benign                                  not provided   \n",
       "7340                  Benign                                  not provided   \n",
       "\n",
       "     OriginSimple                         ReviewStatus Assembly Chromosome  \\\n",
       "1        germline  criteria provided, single submitter   GRCh37         12   \n",
       "3        germline  criteria provided, single submitter   GRCh37         13   \n",
       "5        germline  criteria provided, single submitter   GRCh37          3   \n",
       "9        germline  criteria provided, single submitter   GRCh37          5   \n",
       "11       germline             reviewed by expert panel   GRCh37          5   \n",
       "...           ...                                  ...      ...        ...   \n",
       "7332     germline  criteria provided, single submitter   GRCh37          2   \n",
       "7334     germline  criteria provided, single submitter   GRCh37          2   \n",
       "7336     germline  criteria provided, single submitter   GRCh37          2   \n",
       "7338     germline  criteria provided, single submitter   GRCh37          2   \n",
       "7340     germline  criteria provided, single submitter   GRCh37          7   \n",
       "\n",
       "      PositionVCF ReferenceAlleleVCF AlternateAlleleVCF  RS# (dbSNP)  \\\n",
       "1       123186880                  C                  T      2454727   \n",
       "3        97639879                 GA                  G    145685926   \n",
       "5       151598526                  C                  T     79888638   \n",
       "9        63258565                  C                  C         6295   \n",
       "11       63258565                  C                  G         6295   \n",
       "...           ...                ...                ...          ...   \n",
       "7332    219128291                 CT                  C    775667051   \n",
       "7334    219128334                  A                  G    115448647   \n",
       "7336    219127468                  C                  G    201145373   \n",
       "7338    219127564                  C                  T    770486496   \n",
       "7340      1131378                  C                  T    117290655   \n",
       "\n",
       "                         PhenotypeIDS  \n",
       "1                     MedGen:CN517202  \n",
       "3                     MedGen:CN517202  \n",
       "5                     MedGen:CN517202  \n",
       "9                     MedGen:CN169374  \n",
       "11    MedGen:CN236573|MedGen:CN169374  \n",
       "...                               ...  \n",
       "7332                  MedGen:CN517202  \n",
       "7334                  MedGen:CN517202  \n",
       "7336                  MedGen:CN517202  \n",
       "7338                  MedGen:CN517202  \n",
       "7340                  MedGen:CN517202  \n",
       "\n",
       "[2760 rows x 14 columns]"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_clinvar_classA_gpcr_variants = clinvar_gpcr_variants[\n",
    "    (clinvar_gpcr_variants['Class']=='A') & \\\n",
    "    (clinvar_gpcr_variants['Assembly']=='GRCh37') & \\\n",
    "#    (clinvar_gpcr_variants['ReferenceAlleleVCF'].apply(len) == 1) & \\\n",
    "    (clinvar_gpcr_variants['OriginSimple']=='germline') & \\\n",
    "    (clinvar_gpcr_variants['ReviewStatus'] != 'no assertion criteria provided')\n",
    "]\n",
    "filtered_clinvar_classA_gpcr_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Uncertain significance                                       982\n",
       "Benign                                                       772\n",
       "Likely benign                                                537\n",
       "Conflicting interpretations of pathogenicity                 154\n",
       "Pathogenic                                                    98\n",
       "Likely pathogenic                                             92\n",
       "Benign/Likely benign                                          78\n",
       "Pathogenic/Likely pathogenic                                  25\n",
       "drug response                                                 12\n",
       "not provided                                                   9\n",
       "Conflicting interpretations of pathogenicity, risk factor      1\n",
       "Name: ClinicalSignificance, dtype: int64"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_clinvar_classA_gpcr_variants.ClinicalSignificance.value_counts()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:variants]",
   "language": "python",
   "name": "conda-env-variants-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
